product_default_image/models/product_template.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+from odoo import api, fields, models, tools
+
+
+class ProductTemplate(models.Model):
+
+    _inherit = 'product.template'
+
+    prevent_categ_change_image = fields.Boolean(
+        help='Used to prevent automatic changes to image fields when categ_id '
+        'changes after products have been created.'
+        '(True prevents change)',
+    )
+
+    @api.model
+    def create(self, vals):
+        vals['prevent_categ_change_image'] = True
+        return super(ProductTemplate, self).create(vals)
+
+    @api.model
+    def default_get(self, vals):
+        vals = super(ProductTemplate, self).default_get(vals)
+
+        if vals.get('categ_id'):
+
+            company = self.env.user.company_id
+            category_mod = self.env['product.category']
+            target = company.product_image_target
+
+            if target == 'category':
+                categ = category_mod.browse(vals['categ_id'])
+                vals['image'] = categ.image
+                vals['image_medium'] = categ.image_medium
+                vals['image_small'] = categ.image_small
+
+            elif target == 'global':
+                vals['image'] = company.product_default_global_image
+                tools.image_resize_images(vals)
+
+        return vals
+
+    @api.onchange('categ_id')
+    def _change_image_onchange_categ_id(self):
+        """ Checks if the user has uploaded a custom image or not. """
+
+        company = self.env.user.company_id
+        target = company.product_image_target
+
+        if target == 'category' and not self.prevent_categ_change_image:
+
+            categ_mod = self.env['product.category']
+            categories = categ_mod.search([])
+
+            image_fields = {
+                'image': self.image,
+                'image_medium': self.image_medium,
+                'image_small': self.image_small,
+            }
+
+            matched_categs = []
+            empty_image_field = []
+
+            for key, value in image_fields.iteritems():
```

Comment:
This block is matching up any of the image fields of the product with any category image fields.

If there's a match, it is assumed the user has not uploaded a custom image, and when they change the category, the image will change to the new category's image.

This block also detects if the product's image, image_medium, and image_small fields are None, in which case the image will also change.